sklearn/impute/_knn.py

```diff
@@ -217,71 +217,81 @@ def transform(self, X):
 
         mask = _get_mask(X, self.missing_values)
         mask_fit_X = self._mask_fit_X
+        valid_mask = ~np.all(mask_fit_X, axis=0)
 
-        # Removes columns where the training data is all nan
         if not np.any(mask):
-            valid_mask = ~np.all(mask_fit_X, axis=0)
+            # No missing values in X
+            # Remove columns where the training data is all nan
             return X[:, valid_mask]
 
         row_missing_idx = np.flatnonzero(mask.any(axis=1))
 
-        # Pairwise distances between receivers and fitted samples
-        dist = pairwise_distances(X[row_missing_idx, :], self._fit_X,
-                                  metric=self.metric,
-                                  missing_values=self.missing_values,
-                                  force_all_finite=force_all_finite)
+        non_missing_fix_X = np.logical_not(mask_fit_X)
 
         # Maps from indices from X to indices in dist matrix
         dist_idx_map = np.zeros(X.shape[0], dtype=np.int)
         dist_idx_map[row_missing_idx] = np.arange(row_missing_idx.shape[0])
 
-        non_missing_fix_X = np.logical_not(mask_fit_X)
-
-        # Find and impute missing
-        valid_idx = []
-        for col in range(X.shape[1]):
-
-            potential_donors_idx = np.flatnonzero(non_missing_fix_X[:, col])
-
-            # column was all missing during training
-            if len(potential_donors_idx) == 0:
-                continue
-
-            # column has no missing values
-            if not np.any(mask[:, col]):
-                valid_idx.append(col)
-                continue
+        def process_chunk(dist_chunk, start):
+            row_missing_chunk = row_missing_idx[start:start + len(dist_chunk)]
 
-            valid_idx.append(col)
-
-            receivers_idx = np.flatnonzero(mask[:, col])
-
-            # distances for samples that needed imputation for column
-            dist_subset = (dist[dist_idx_map[receivers_idx]]
-                           [:, potential_donors_idx])
+            # Find and impute missing by column
+            for col in range(X.shape[1]):
+                if not valid_mask[col]:
+                    # column was all missing during training
+                    continue
 
-            # receivers with all nan distances impute with mean
-            all_nan_dist_mask = np.isnan(dist_subset).all(axis=1)
-            all_nan_receivers_idx = receivers_idx[all_nan_dist_mask]
+                col_mask = mask[row_missing_chunk, col]
+                if not np.any(col_mask):
+                    # column has no missing values
+                    continue
 
-            if all_nan_receivers_idx.size:
-                col_mean = np.ma.array(self._fit_X[:, col],
-                                       mask=mask_fit_X[:, col]).mean()
-                X[all_nan_receivers_idx, col] = col_mean
+                potential_donors_idx, = np.nonzero(non_missing_fix_X[:, col])
```

Comment:
maybe use `np.flatnonzero` or is there any efficiency to use `np.nonzero` instead?

Reply:
Only reason is line length :\

Reply:
Oh ok, why not :)